3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/adapters/registry/VeroAdapter.scala

```diff
@@ -0,0 +1,167 @@
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics
+package snowplow
+package enrich
+package common
+package adapters
+package registry
+
+// Java
+import com.fasterxml.jackson.core.JsonParseException
+
+// Scalaz
+import scalaz.Scalaz._
+
+// json4s
+import org.json4s._
+import org.json4s.jackson.JsonMethods._
+
+// Iglu
+import com.snowplowanalytics.iglu.client.{Resolver, SchemaKey}
+
+// Joda Time
+import org.joda.time.{DateTime, DateTimeZone}
+import org.joda.time.format.{DateTimeFormat, DateTimeFormatter}
+
+// This project
+import com.snowplowanalytics.snowplow.enrich.common.loaders.CollectorPayload
+//import com.snowplowanalytics.snowplow.enrich.common.utils.{JsonUtils => JU}
+
+import scala.util.{Failure, Success, Try}
+
+/**
+ * Transforms a collector payload which conforms to
+ * a known version of the Vero webhook
+ * into raw events.
+ */
+object VeroAdapter extends Adapter {
+
+  // Vendor name for Failure Message
+  private val VendorName = "Vero"
+
+  // Expected content type for a request body
+  private val ContentType = "application/json"
+
+  // Tracker version for an Vero webhook
+  private val TrackerVersion = "com.getvero-v1"
+
+  // Schemas for reverse-engineering a Snowplow unstructured event
+  private val EventSchemaMap = Map(
+    "bounced"      -> SchemaKey("com.getvero", "bounced", "jsonschema", "1-0-0").toSchemaUri,
+    "clicked"      -> SchemaKey("com.getvero", "clicked", "jsonschema", "1-0-0").toSchemaUri,
+    "delivered"    -> SchemaKey("com.getvero", "delivered", "jsonschema", "1-0-0").toSchemaUri,
+    "opened"       -> SchemaKey("com.getvero", "opened", "jsonschema", "1-0-0").toSchemaUri,
+    "sent"         -> SchemaKey("com.getvero", "sent", "jsonschema", "1-0-0").toSchemaUri,
+    "unsubscribed" -> SchemaKey("com.getvero", "unsubscribed", "jsonschema", "1-0-0").toSchemaUri,
+    "user_created" -> SchemaKey("com.getvero", "created", "jsonschema", "1-0-0").toSchemaUri,
+    "user_updated" -> SchemaKey("com.getvero", "updated", "jsonschema", "1-0-0").toSchemaUri
+  )
+
+  /**
+   * Converts a payload into a single validated event
+   * Expects a valid json returns failure if one is not present
+   * For "user_updated" events the field is named "action" instead of "type"
+   * in which case the event type will be set to Some("user_updated")
+   *
+   * @param json Payload body that is sent by Vero
+   * @param payload The details of the payload
+   * @return a Validation boxing either a NEL of RawEvents on
+   *         Success, or a NEL of Failure Strings
+   */
+  private def payloadBodyToEvent(json: String, payload: CollectorPayload): Validated[RawEvent] = {
+
+    val parsed = Try(parse(json))
+    parsed match {
+      case Success(parsed) => parsed
+      case Failure(e)      => return s"$VendorName event failed to parse into JSON: [${e.getMessage}]".failureNel
+    }
```

Comment:
the result of the matchmaking is discarded, it should be:

```scala
val parsed = Try(parse(json)) match {
  case Success(parsed) => parsed
  case Failure(e) => s"$VendorName event failed to parse into JSON: [${e.getMessage}]".failureNel
}
```

Reply:
Heyo @BenFradet , I found that if I don't "return" early here on failure then running the test specs spits out an error instead of a failure.
E.g. passing in a bad json payload "hello" when tested shows:
```
! toRawEvents must return a success for a valid for all event types payload body being passed
[error]  JsonParseException: : Unrecognized token 'hello': was expecting ('true', 'false' or 'null')
[error]  at [Source: java.io.StringReader@42b5aa30; line: 1, column: 11]  (JsonParser.java:1524)
```
versus
```
x toRawEvents must return a success for a valid for all event types payload body being passed
[error]  'Failure(NonEmptyList(Twilio event failed to parse into JSON: [Unrecognized token 'hello': was expecting ('true', 'false' or 'null')
[error]   at [Source: java.io.StringReader@38849b9f; line: 1, column: 11]]))' is not Success with value'NonEmptyList(...
```

Reply:
you don't need to return early you just need to carry on with your failure, this is why there is a `parsed.map` in my comment below because `parsed` will have type `ValidationNel[String, JValue]`.

by the way, the comment above should say `case Success(parsed) => parsed.successNel`, that's my bad.